File: ui/src/api/progress.ts
// Progress & Recommendations API functions
import { api } from './client';
import type { Module } from './modules';

export interface ProgressSummary {
    child_id: string;
    total_time_minutes: number;
    total_modules_completed: number;
    total_questions_answered: number;
    average_accuracy: number;
    current_streak_days: number;
    total_points: number;
    favorite_module_type?: string;
    most_active_time?: string;
}

export interface ProgressEventCreate {
    module_id: string;
    question_id?: string;
    is_correct: boolean;
    time_taken_seconds: number;
    attempt_count?: number;
}

export interface ProgressResponse {
    id: string;
    child_id: string;
    module_id: string;
    question_id?: string;
    is_correct: boolean;
    time_taken_seconds: number;
    attempt_count: number;
    points_earned: number;
    created_at: string;
}

export interface RecommendedModule {
    module: Module;
    confidence_score: number;
    reasons: { factor: string; weight: number; description: string }[];
    expected_difficulty: number;
}

export interface RecommendationResponse {
    child_id: string;
    recommended_modules: RecommendedModule[];
    next_best_module: RecommendedModule;
    personalization_level: string;
    generated_at: string;
    valid_until: string;
}

export const progressApi = {
    getSummary: async (childId: string, days = 30): Promise<ProgressSummary> => {
        return api.get<ProgressSummary>(`/progress/children/${childId}/summary?days=${days}`);
    },

    saveProgress: async (childId: string, data: ProgressEventCreate): Promise<ProgressResponse> => {
        return api.post<ProgressResponse>(`/progress/children/${childId}/events`, data);
    },
};

export const recommendationsApi = {
    get: async (childId: string): Promise<RecommendationResponse> => {
        return api.get<RecommendationResponse>(`/recommendations/children/${childId}`);
    },

    getNextModule: async (childId: string): Promise<RecommendedModule> => {
        return api.get<RecommendedModule>(`/recommendations/children/${childId}/next-module`);
    },
};

File: app/schemas/progress.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict
from datetime import datetime


class ProgressEventCreate(BaseModel):
    """Schema for creating a progress event"""
    module_id: str
    question_id: Optional[str] = None
    is_correct: bool
    time_taken_seconds: int = Field(..., ge=0)
    attempt_count: int = Field(default=1, ge=1)


class ProgressBatchCreate(BaseModel):
    """Schema for batch progress sync"""
    events: List[ProgressEventCreate]
    offline_session_id: str
    synced_at: datetime = Field(default_factory=datetime.utcnow)


class ProgressResponse(BaseModel):
    """Schema for progress response"""
    id: str
    child_id: str
    module_id: str
    question_id: Optional[str] = None
    is_correct: bool
    time_taken_seconds: int
    attempt_count: int
    points_earned: int
    created_at: datetime


class ProgressSummary(BaseModel):
    """Schema for progress summary"""
    child_id: str
    total_time_minutes: int
    total_modules_completed: int
    total_questions_answered: int
    average_accuracy: float
    current_streak_days: int
    total_points: int
    favorite_module_type: Optional[str] = None
    most_active_time: Optional[str] = None


class SubjectProgress(BaseModel):
    """Schema for subject-specific progress"""
    subject: str
    accuracy: float
    total_questions: int
    time_spent_minutes: int
    level: int
    progress_percentage: float


class StrengthWeakness(BaseModel):
    """Schema for strength/weakness analysis"""
    category: str
    skill_name: str
    performance_score: float  # 0-100
    recommendation: str


class ProgressReport(BaseModel):
    """Schema for detailed progress report"""
    child_id: str
    period_start: datetime
    period_end: datetime
    overall_summary: ProgressSummary
    subject_progress: List[SubjectProgress]
    strengths: List[StrengthWeakness]
    areas_for_improvement: List[StrengthWeakness]
    recent_achievements: List[str]
    weekly_activity: Dict[str, int]  # day -> minutes
    generated_at: datetime